File: cmd/hubctl/Client_test.go
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2026 Jérôme Bastian Winkel

package main

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"

	"github.com/Jersyfi/hubtask/core/shared/secret"
	"github.com/Jersyfi/hubtask/infrastructure/i18n"
)

func catalogue(t *testing.T) i18n.Catalogue {
	t.Helper()
	loaded, err := i18n.LoadEnglish()
	if err != nil {
		t.Fatalf("loading the catalogue: %v", err)
	}
	return loaded
}

// installation stands in for a Hubtask installation. It records the last request, so that a test
// can assert on what the client sent as well as on what it made of the answer.
type installation struct {
	server  *httptest.Server
	request *http.Request
	body    string
}

func serve(t *testing.T, handler http.HandlerFunc) *installation {
	t.Helper()
	stub := &installation{}
	stub.server = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		stub.request = r.Clone(r.Context())
		if r.Body != nil {
			raw, _ := io.ReadAll(r.Body)
			stub.body = string(raw)
		}
		handler(w, r)
	}))
	t.Cleanup(stub.server.Close)
	return stub
}

func clientFor(t *testing.T, stub *installation) *Client {
	t.Helper()
	client, err := NewClient(
		Profile{BaseURL: stub.server.URL, Token: secret.New("hbt_pat_token")},
		catalogue(t), 5*time.Second)
	if err != nil {
		t.Fatalf("building the client: %v", err)
	}
	return client
}

// problemJSON writes an RFC 9457 document the way presentation/rest does.
func problemJSON(w http.ResponseWriter, status int, document map[string]any) {
	w.Header().Set("Content-Type", "application/problem+json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(document)
}

func TestTheCredentialTravelsInTheHeaderAndThePathCarriesTheAPIVersion(t *testing.T) {
	stub := serve(t, func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"data":[]}`))
	})

	var payload struct{ Data []any }
	if err := clientFor(t, stub).Get(context.Background(), "/containers", nil, &payload); err != nil {
		t.Fatalf("calling: %v", err)
	}

	if got := stub.request.Header.Get("Authorization"); got != "Bearer hbt_pat_token" {
		t.Errorf("Authorization %q", got)
	}
	if got := stub.request.URL.Path; got != APIPath+"/containers" {
		t.Errorf("path %q, want %q", got, APIPath+"/containers")
	}
	// The token belongs in the header and nowhere else: a URL travels into logs and into the
	// Referer of whatever the target renders.
	if strings.Contains(stub.request.URL.String(), "hbt_pat") {
		t.Error("the token reached the URL")
	}
}

func TestARefusalIsRenderedFromTheCatalogue(t *testing.T) {
	for _, tc := range []struct {
		name     string
		document map[string]any
		status   int
		want     string
	}{
		{
			name:     "the detail code wins where the catalogue knows it",
			status:   http.StatusNotFound,
			document: map[string]any{"status": 404, "code": "not_found", "detail_code": "containers.parent_not_found"},
			want:     "containers.parent_not_found",
		},
		{
			name:     "the contract code carries an answer with no detail code",
			status:   http.StatusNotFound,
			document: map[string]any{"status": 404, "code": "not_found"},
			want:     "errors.not_found",
		},
		{
			name:     "a detail code the catalogue has never heard of falls back to the contract code",
			status:   http.StatusConflict,
			document: map[string]any{"status": 409, "code": "conflict", "detail_code": "items.invented_for_this_test"},
			want:     "errors.conflict",
		},
	} {
		t.Run(tc.name, func(t *testing.T) {
			stub := serve(t, func(w http.ResponseWriter, _ *http.Request) {
				problemJSON(w, tc.status, tc.document)
			})

			err := clientFor(t, stub).Get(context.Background(), "/containers", nil, nil)
			if err == nil {
				t.Fatal("a refusal came back as a success")
			}

			expected, known := catalogue(t).Message(tc.want, nil)
			if !known {
				t.Fatalf("the test expects %s to be in the catalogue", tc.want)
			}
			if err.Error() != expected {
				t.Errorf("message %q, want %q", err.Error(), expected)
			}
			// Whatever else is true of the message, it is not the document.
			if strings.Contains(err.Error(), "{") {
				t.Errorf("the message %q looks like JSON", err.Error())
			}
		})
	}
}

// The parameters are what makes a message a sentence rather than a template.
func TestTheParametersOfAProblemReachTheMessage(t *testing.T) {
	stub := serve(t, func(w http.ResponseWriter, _ *http.Request) {
		problemJSON(w, http.StatusInternalServerError, map[string]any{
			"status": 500, "code": "internal", "request_id": "01J8ZQ",
		})
	})

	err := clientFor(t, stub).Get(context.Background(), "/containers", nil, nil)
	if err == nil {
		t.Fatal("a 500 came back as a success")
	}
	if !strings.Contains(err.Error(), "01J8ZQ") {
		t.Errorf("the message %q does not quote the request ID the user is asked to reference", err.Error())
	}
}

func TestFieldErrorsAreRenderedUnderTheMessage(t *testing.T) {
	stub := serve(t, func(w http.ResponseWriter, _ *http.Request) {
		problemJSON(w, http.StatusUnprocessableEntity, map[string]any{
			"status": 422, "code": "validation_failed",
			"field_errors": []map[string]any{
				{"path": "/name", "code": "shared.required"},
			},
		})
	})

	err := clientFor(t, stub).Get(context.Background(), "/containers", nil, nil)
	if err == nil {
		t.Fatal("a validation failure came back as a success")
	}
	lines := strings.Split(err.Error(), "\n")
	if len(lines) != 2 {
		t.Fatalf("%d lines, want the message and one field: %q", len(lines), err.Error())
	}
	if !strings.Contains(lines[1], "name:") {
		t.Errorf("the field line %q does not name the field", lines[1])
	}
}

// The server puts the parameters on the problem and the code on the field, so a field error
// rendered from its own parameters alone would print its placeholders at a user. This is the case
// the end-to-end session caught: "A {item_type} cannot sit in a {parent_type}."
func TestAFieldErrorBorrowsTheProblemsParameters(t *testing.T) {
	stub := serve(t, func(w http.ResponseWriter, _ *http.Request) {
		problemJSON(w, http.StatusUnprocessableEntity, map[string]any{
			"status": 422, "code": "capability_not_supported",
			"detail_code": "items.parent_type_invalid",
			"params":      map[string]any{"item_type": "ACTIVITY", "parent_type": "TASK"},
			"field_errors": []map[string]any{
				{"path": "/parent_id", "code": "items.parent_item_required"},
			},
		})
	})

	err := clientFor(t, stub).Get(context.Background(), "/items", nil, nil)
	if err == nil {
		t.Fatal("a validation failure came back as a success")
	}
	if strings.Contains(err.Error(), "{") {
		t.Errorf("a placeholder reached the message: %q", err.Error())
	}
	// The field line is rendered from the problem's parameters, which is where the server puts
	// them - so it names the type rather than printing the placeholder for it.
	if !strings.Contains(err.Error(), "parent_id: ") || !strings.Contains(err.Error(), "ACTIVITY") {
		t.Errorf("the field line is missing or empty: %q", err.Error())
	}
}

// Two identical sentences read as a bug in the client rather than as detail: the field line adds
// only the field's name, and the sentence above it has already named the problem.
func TestAFieldErrorThatRepeatsTheMessageIsNotPrintedTwice(t *testing.T) {
	stub := serve(t, func(w http.ResponseWriter, _ *http.Request) {
		problemJSON(w, http.StatusNotFound, map[string]any{
			"status": 404, "code": "not_found", "detail_code": "items.collection_not_found",
			"field_errors": []map[string]any{
				{"path": "/collection_id", "code": "items.collection_not_found"},
			},
		})
	})

	err := clientFor(t, stub).Get(context.Background(), "/items", nil, nil)
	if err == nil {
		t.Fatal("a refusal came back as a success")
	}
	if strings.Contains(err.Error(), "\n") {
		t.Errorf("the message was printed twice: %q", err.Error())
	}
}

// A proxy, a captive portal or a load balancer answers with HTML. Echoing that at a user would
// be printing somebody else's page into their terminal.
func TestAnAnswerThatIsNotAProblemDocumentIsReportedByItsStatus(t *testing.T) {
	stub := serve(t, func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "text/html")
		w.WriteHeader(http.StatusBadGateway)
		_, _ = w.Write([]byte("<html><body>Gateway problem</body></html>"))
	})

	err := clientFor(t, stub).Get(context.Background(), "/containers", nil, nil)
	if err == nil {
		t.Fatal("a 502 came back as a success")
	}
	if !strings.Contains(err.Error(), "502") {
		t.Errorf("the message %q does not name the status", err.Error())
	}
	if strings.Contains(err.Error(), "Gateway problem") {
		t.Errorf("the body reached the message: %q", err.Error())
	}
}

func TestAnEmptyAnswerIsASuccess(t *testing.T) {
	stub := serve(t, func(w http.ResponseWriter, _ *http.Request) {
		w.WriteHeader(http.StatusNoContent)
	})

	if err := clientFor(t, stub).Delete(context.Background(), "/items/x", `W/"3"`); err != nil {
		t.Fatalf("deleting: %v", err)
	}
	if got := stub.request.Header.Get("If-Match"); got != `W/"3"` {
		t.Errorf("If-Match %q", got)
	}
}

func TestABodyIsSentAsJSON(t *testing.T) {
	stub := serve(t, func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusCreated)
		_, _ = w.Write([]byte(`{"name":"Hub"}`))
	})

	var created struct{ Name string }
	if err := clientFor(t, stub).Post(context.Background(), "/containers",
		map[string]string{"name": "Hub"}, &created); err != nil {
		t.Fatalf("creating: %v", err)
	}
	if stub.body != `{"name":"Hub"}` {
		t.Errorf("body %q", stub.body)
	}
	if got := stub.request.Header.Get("Content-Type"); got != "application/json" {
		t.Errorf("Content-Type %q", got)
	}
	if created.Name != "Hub" {
		t.Errorf("the answer was not decoded: %+v", created)
	}
}

func TestAClientNeedsAnAddressAndACredential(t *testing.T) {
	for _, tc := range []struct {
		name    string
		profile Profile
		want    string
	}{
		{"no address", Profile{Token: secret.New("t")}, "no installation"},
		{"no credential", Profile{BaseURL: "https://hub.example.com"}, "not signed in"},
	} {
		t.Run(tc.name, func(t *testing.T) {
			_, err := NewClient(tc.profile, catalogue(t), time.Second)
			if err == nil {
				t.Fatal("a client was built without what it needs")
			}
			if !strings.Contains(err.Error(), tc.want) {
				t.Errorf("message %q, want it to contain %q", err.Error(), tc.want)
			}
		})
	}
}

// The guard's own refusals are message codes too, and they must not read as though they came
// from a different program than the server's.
func TestAnAddressTheGuardRefusesIsRenderedFromTheCatalogue(t *testing.T) {
	client, err := NewClient(
		Profile{BaseURL: "https://hub.invalid", Token: secret.New("t")}, catalogue(t), 2*time.Second)
	if err != nil {
		t.Fatalf("building the client: %v", err)
	}

	err = client.Get(context.Background(), "/containers", nil, nil)
	if err == nil {
		t.Fatal("a call to a name that does not resolve succeeded")
	}
	if strings.Contains(err.Error(), "dependency.") || strings.Contains(err.Error(), "errors.") {
		t.Errorf("the message %q is a code rather than a sentence", err.Error())
	}
}

File: sdk/typescript/scripts/client.test.js
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2026 Jérôme Bastian Winkel

// The generated TypeScript client against a server that records what it received (P-03). The
// typing is proved by `pnpm typecheck` against dist/schema.d.ts; this proves the runtime half -
// the path filled, the query appended, the bearer and the headers sent, the body encoded, and a
// refusal decoded into ProblemError.

import assert from 'node:assert/strict';
import http from 'node:http';
import test from 'node:test';

import { HubtaskClient, ProblemError } from '../src/client.gen.ts';

async function withServer(handler, run) {
  const received = [];
  const server = http.createServer((request, response) => {
    let body = '';
    request.on('data', (chunk) => (body += chunk));
    request.on('end', () => {
      received.push({ method: request.method, url: request.url, headers: request.headers, body });
      handler(request, response, body);
    });
  });
  await new Promise((resolve) => server.listen(0, '127.0.0.1', resolve));
  const { port } = server.address();
  try {
    await run(`http://127.0.0.1:${port}/api/v1`, received);
  } finally {
    server.close();
  }
}

test('a read fills the path, appends the query and sends the bearer', async () => {
  await withServer(
    (_, response) => {
      response.setHeader('Content-Type', 'application/json');
      response.end(JSON.stringify({ data: [{ id: 'c1', name: 'Private' }], page: { next_cursor: null, has_more: false } }));
    },
    async (baseUrl, received) => {
      const client = new HubtaskClient({ baseUrl, token: () => 'hbt_pat_x', headers: { 'User-Agent': 'test' } });
      const page = await client.listContainers({ query: { type: 'COLLECTION', parent_id: 'h 1', size: 5 } });
      assert.equal(page.data[0].name, 'Private');
      const [call] = received;
      assert.equal(call.method, 'GET');
      assert.equal(call.url, '/api/v1/containers?type=COLLECTION&parent_id=h+1&size=5');
      assert.equal(call.headers.authorization, 'Bearer hbt_pat_x');
      assert.equal(call.headers['user-agent'], 'test');
      assert.match(call.headers.accept, /application\/json/);
    },
  );
});

test('a mutation encodes the body and sends the idempotency key; a path parameter is escaped', async () => {
  await withServer(
    (_, response) => {
      response.statusCode = 201;
      response.setHeader('Content-Type', 'application/json');
      response.end(JSON.stringify({ id: 'i1', title: 'Buy oat milk', version: 1 }));
    },
    async (baseUrl, received) => {
      const client = new HubtaskClient({ baseUrl, token: 'hbt_pat_x' });
      const created = await client.createWorkItem({ type: 'TASK', title: 'Buy oat milk' }, { idempotencyKey: 'k-1' });
      assert.equal(created.id, 'i1');
      await client.updateWorkItem('a/b', { title: 'Renamed' }, { ifMatch: '"1"' });
      const [create, update] = received;
      assert.equal(create.headers['content-type'], 'application/json');
      assert.equal(create.headers['idempotency-key'], 'k-1');
      assert.deepEqual(JSON.parse(create.body), { type: 'TASK', title: 'Buy oat milk' });
      assert.equal(update.method, 'PATCH');
      assert.equal(update.url, '/api/v1/items/a%2Fb');
      assert.equal(update.headers['content-type'], 'application/merge-patch+json');
      assert.equal(update.headers['if-match'], '"1"');
    },
  );
});

test('a refusal is a ProblemError with the document, and a body that is not one still has the status', async () => {
  await withServer(
    (request, response) => {
      if (request.url.endsWith('/meta/health')) {
        response.statusCode = 502;
        response.end('<html>upstream</html>');
        return;
      }
      response.statusCode = 422;
      response.setHeader('Content-Type', 'application/problem+json');
      response.end(JSON.stringify({ status: 422, code: 'validation_failed', detail_code: 'items.title_too_long', field_errors: [{ path: '/title' }] }));
    },
    async (baseUrl) => {
      const client = new HubtaskClient({ baseUrl });
      await assert.rejects(client.createWorkItem({ type: 'TASK', title: '' }), (error) => {
        assert.ok(error instanceof ProblemError);
        assert.equal(error.status, 422);
        assert.equal(error.problem.code, 'validation_failed');
        assert.equal(error.problem.field_errors[0].path, '/title');
        assert.equal(error.message, 'hubtask: 422 validation_failed (items.title_too_long)');
        return true;
      });
      await assert.rejects(client.getHealthReport(), (error) => {
        assert.ok(error instanceof ProblemError);
        assert.equal(error.status, 502);
        assert.equal(error.problem.code, '');
        return true;
      });
    },
  );
});

test('a 204 answers nothing and a raw answer is the response itself', async () => {
  await withServer(
    (request, response) => {
      if (request.method === 'DELETE') {
        response.statusCode = 204;
        response.end();
        return;
      }
      response.setHeader('Content-Type', 'text/calendar');
      response.end('BEGIN:VCALENDAR');
    },
    async (baseUrl, received) => {
      const client = new HubtaskClient({ baseUrl, token: 'hbt_pat_x' });
      assert.equal(await client.trashWorkItem('i1', { ifMatch: '"2"' }), undefined);
      const feed = await client.getCalendarFeedDocument('tok');
      assert.equal(await feed.text(), 'BEGIN:VCALENDAR');
      assert.doesNotMatch(received[1].headers.accept ?? '', /application\/json/, 'a raw answer does not ask for JSON');
    },
  );
});

test('the quickstart example walks a hub against a recording server', async () => {
  const { run } = await import('../examples/quickstart.mjs');
  await withServer(
    (request, response) => {
      response.setHeader('Content-Type', 'application/json');
      if (request.url.startsWith('/api/v1/containers')) {
        response.end(JSON.stringify({ data: [{ id: 'c1', name: 'Inbox' }], page: { next_cursor: null, has_more: false } }));
      } else if (request.method === 'POST') {
        response.statusCode = 201;
        response.end(JSON.stringify({ id: 'i1', title: 'Made by the TypeScript SDK', version: 1 }));
      } else {
        response.end(JSON.stringify({ id: 'i1', title: 'Made by the TypeScript SDK', version: 1 }));
      }
    },
    async (baseUrl, received) => {
      const lines = [];
      const read = await run({ baseUrl, token: 'hbt_pat_x', hub: 'h1', log: (line) => lines.push(line) });
      assert.equal(read.title, 'Made by the TypeScript SDK');
      assert.deepEqual(received.map((call) => `${call.method} ${call.url}`), [
        'GET /api/v1/containers?type=COLLECTION&parent_id=h1',
        'POST /api/v1/items',
        'GET /api/v1/items/i1',
      ]);
      assert.equal(lines.length, 3);
    },
  );
});
